packages/mui-system/src/createStyled/createStyled.js

```diff
@@ -14,6 +14,53 @@ export function shouldForwardProp(prop) {
   return prop !== 'ownerState' && prop !== 'theme' && prop !== 'sx' && prop !== 'as';
 }
 
+function pure(styleFn) {
+  let lastTheme;
+  let lastValue;
+
+  return (props) => {
+    const theme = props.theme;
+    let value = lastValue;
+    if (value === undefined || theme !== lastTheme) {
+      value = styleFn(theme);
+
+      // Example preprocessing we can do if `style` is stable.
+      preprocessVariants(value)
+
+      lastValue = value;
+      lastTheme = theme;
+    }
+    return value;
+  };
+}
+
+function preprocessVariants(style) {
+  // Compiles predicate functions for each variant, to add `variant.matches(props)`.
```

Comment:
This is an example of optimization that can be unlocked once we know the `style` is static & shared across all instances of a component. It wouldn't make sense to eval-compile predicates if the style object was changing for each Button for example.